Feature/ckswnskfk/2

ckswnskfk opened the pull request:
## 어떤 변경인지 
- [x] Feat: 기능 변경, 기능 추가
- [ ] Docs: 문서 작업
- [ ] Refactor: 기능 변경 없이 코드 수정
- [ ] Style: 디자인
- [ ] Fix: 버그 수정
- [ ] Test: 테스트 코드 작성

## 설명
Input 공통 컴포넌트를 만들었습니다.

width는 기본적으로 100%로 설정했습니다. 크기 설정이 필요하시면 className으로 넘겨주시면 될 것 같습니다!
errorMessage prop을 넘기면 조건부 렌더링으로 에러 메시지가 출력됩니다.

- 리뷰 반영
1. clsx 사용으로 가독성 증진
2. focus 시 레이아웃 변경으 일으키지 않기 위해 border 대신 box-shadow 로 변경
3. disabled는 pseudo class로 넘겨서 처리함으로 깔끔한 코드 유지
4. errorMessage 조건부 렌더링 시 앞에 !! 추가로 boolean 타입으로 만들어 안정성 확보

### 이슈 번호
https://github.com/Codeit-Rolling-11-Letsgo/Rolling/issues/2

### 주요 변경 사항 

### 어떻게 동작하는지

### To Reviewers
시안을 보면 input은 label과 같이 사용되는데 아예 label이 포함된 InputWithLabel 이런 컴포넌트를 만드는 건 어떨까요?
그렇지 않으면 id를 넘겨줘야해서 번거로울 수 있는데 컴포넌트 내부에서 useId 같은 걸 사용해 한 번에 처리할 수도 있을 것 같습니다. 다만 시안에는 그럴 일이 없지만 label의 스타일을 지정해야 할 땐 prop이 좀 복잡해 보일 수도 있겠네요. 의견 주세요!

ckswnskfk committed:
Feat: Input 공통 컴포넌트 추가

ckswnskfk committed:
Feat: Input 컴포넌트 jsdoc 추가

ckswnskfk committed:
Style: 누락된 에러 메시지 간격 추가

ckswnskfk commented on src/components/common/Input.jsx:
```diff
@@ -0,0 +1,25 @@
+
+/**
+ *
+ * @param {{errorMessage?: string}} props
```
jsdoc은 좀 더 알아봐야겠지만 컨벤션으로 정해도 좋을 것 같네요!

mizign commented on src/components/common/Input.jsx:
```diff
@@ -0,0 +1,25 @@
+function Input({ errorMessage, className, ...props }) {
+	console.log(props);
+	const { disabled } = { ...props };
+	const classNames = `${styles.input} ${disabled && styles.disabled} ${
```
오 disabled 어떻게 할지 고민중이었는데,, 이러면 되겠네요! 퍼갑니다

ckswnskfk committed:
Style: Input 폴더 생성 및 관련 파일 이동

mizign commented on src/components/common/Input/Input.jsx:
```diff
@@ -0,0 +1,25 @@
+ * @param {{errorMessage?: string}} props
+ * @returns
+ */
+function Input({ errorMessage, className, ...props }) {
+	console.log(props);
+	const { disabled } = { ...props };
```
errorMessage는 prop으로 명시하시고 disabled 프롭은 ...props로 구조 분해 하신 의도가 궁금해요! 둘 다 매번 오는 prop은 아니니까 둘 다 ...prop에서 구조분해 해줘도 되지 않나요?

ckswnskfk replied:
아래에서 disabled 에 따라 클래스를 지정하기 위해 이렇게 빼서 사용해봤습니다!
근데 console.log가 들어가 있네요.. 얼른 빼겠습니다ㅋㅋㅋ

dkile replied:
콘솔 로그도 husky에서 push 전에 잡을 수 있도록 해둘게요

ckswnskfk committed:
Refactor: console.log 삭제

dkile commented on src/components/common/Input/Input.jsx:
```diff
@@ -0,0 +1,24 @@
+ */
+function Input({ errorMessage, className, ...props }) {
+	const { disabled } = { ...props };
+	const classNames = `${styles.input} ${disabled && styles.disabled} ${
```
https://github.com/lukeed/clsx
clsx를 활용해보시는건 어떤가요?

dkile replied:
disabled의 스타일링은 pseudo class로 css에게 맡기고 disabled는 ...props 안에 포함시키면 될 것 같은데 혹시 다른 의도가 있으셨나요?
이 부분은 그냥 궁금한거라서 그런거에요.
딱히 가독성이나 사용성에는 불편함이 없습니다.

ckswnskfk replied:
disabled pseudo 클래스가 있는 줄 몰라서 따로 빼서 조건에 따라 클래스를 넣기 위해 이렇게 사용했네요.. 😢

dkile commented on src/components/common/Input/Input.module.scss:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+.input.disabled {
```
pseudo class를 사용하지 않은 이유가 있으실까요?

dkile commented on src/components/common/Input/Input.module.scss:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+.input.error {
```
:invalid pseudo class를 사용해보는건 혹시 어떤가요?

ckswnskfk replied:
invalid 도 있는 줄 몰랐습니다...!

dkile commented on src/components/common/Input/Input.jsx:
```diff
@@ -0,0 +1,24 @@
+	} ${className}`;
+
+	return (
+		<>
```
바깥쪽을 div로 감싸고 errorMessage와 input 사이의 간격도 공통 컴포넌트 내에서 정의하는게 어떨까요?

ckswnskfk replied:
그 방법도 생각했었는데 사용하는 입장에서 의도치 않은 div가 생긴다고 느낄 것 같아서 그냥 이렇게 처리 했습니다.
별로 상관 없을까요?

dkile replied:
네네 크게 상관 없을 것 같네요

dkile commented:
border-radius나 border도 px을 rem으로 변경하는게 실제로 해보니까 더 자연스럽네요.
한 번 보여드리겠습니다.

d0ngbb00 commented:
저는 아는게 많지 않아서 다른 분들이 남겨주신 리뷰를 토대로 덩달아 같이 공부하고 있네요 ..😂
특히 찬주님 코드를 통해서
```
function Input({ errorMessage, className, ...props }) {
	const { disabled } = { ...props };
```
이런 방법으로 props에서 특정 속성을 추출할 수 있다는 걸 새롭게 배웠습니다 !

ckswnskfk committed:
Refactor: 리뷰 사항 반영

clsx 사용으로 가독성 증진
focus 시 레이아웃 변경으 일으키지 않기 위해 border 대신 box-shadow 로 변경
disabled는 pseudo class로 넘겨서 처리함으로 깔끔한 코드 유지
errorMessage 조건부 렌더링 시 앞에 !! 추가로 boolean 타입으로 만들어 안정성 확보

mizign reviewed (approved):
clsx 저도 써봐야겠네요,, 제가 보기엔 더 고칠게 없는 코드 같아 일단 approve 해둡니다.
